### PR TITLE
Add MARK and SQUASH builder instructions

### PR DESCRIPTION
**Pull request by burke**
## TL;DR

There are actually docs in the diff, but the 10-second overview is:

``` Dockerfile
FROM scratch
MAINTAINER me
MARK
ADD a
RUN b
SQUASH ">> important things <<"
```

```
$ docker history ${id}
IMAGE               CREATED                  CREATED BY                          SIZE
2a1491d233a1        Less than a second ago   >> important things <<              20 KB
df7546f9f060        6 months ago             /bin/sh -c #(nop) MAINTAINER me     0 B
511136ea3c5a        22 months ago                                                0 B
$
```
## Summary

This PR adds two commands -- `MARK` and `SQUASH` -- to the builder. I've wanted some way to implement "transactions", whereby multiple instructions are ultimately reduced to a single layer, in `Dockerfile`s for a long time. This was by far the least invasive way to implement it.
- `MARK` stashes a tiny bit of context on the `Builder` struct so that `SQUASH` can look it up later.
- `SQUASH` squashes the indicated layers between `MARK` and `SQUASH` together into a single new layer. 
- `SQUASH` also gives the new layer a new `ContainerConfig.Cmd`, whose only apparent use is to show up in `docker history`. This can be used to substantially prettify `docker history` output.
## Why is this necessary?
1. At Shopify, some of our containers require secrets to be temporarily decrypted during build time. There are trivially-conceivable three strategies for this. Two of them are impossible currently, and one is sketchy. See Appendix A. `SQUASH` makes it possible to have temporary build-time secrets without them existing in the image hierarchy.
2. It's fairly common to want to `ADD` some files, take some action with them, then to remove them. This causes the final image distribution to be much larger than necessary. If this can be compressed to a single layer, network and disk are used more efficiently.
3. At Shopify, we enjoy having prettified image histories. For example, this is our current core platform production image history:

```
IMAGE               CREATED             CREATED BY                                      SIZE
3a342817ba4a        37 minutes ago      ##build 3 & finalize @d9c88f1c                  x MB
4a4f06aab8d0        9 hours ago         ##build 2 @1d9a653a                             y MB
5e004daccb40        2 days ago          ##build 1 @a1ecb04a                             z MB
38fee11f123e        2 weeks ago         ##prepare                                       w MB
623b6b3205b7        2 weeks ago         ##baseimage v1.3.11                             v MB
d0955f21bf24        2 weeks ago         /bin/sh -c #(nop) CMD [/bin/bash]               0 B
9fec74352904        2 weeks ago         /bin/sh -c sed -i 's/^#\s*\(deb.*universe\)$/   1.895 kB
a1a958a24818        2 weeks ago         /bin/sh -c echo '#!/bin/sh' > /usr/sbin/polic   194.5 kB
f3c84ac3a053        2 weeks ago         /bin/sh -c #(nop) ADD file:777fad733fc954c0c1   188.1 MB
511136ea3c5a        22 months ago                                                       0 B
```

This `MARK`/`SQUASH` would make our rather insane custom build process theoretically-achievable using only a handful of strategically-composed `Dockerfile`s.
## Conclusions...

There's still a lot of testing and polish to be applied here, but before I continue, I want to get a sense of how likely this is to be accepted; i.e. is the docker project still philosophically opposed to acknowledging the existence of layers from the context of a `Dockerfile`?

Does this seem like the kind of that that could possibly move forward?

/cc Shopify peeps: @graemej @Sirupsen @thegedge @shivnagarajan

---
#### Appendix A: Obvious ways to get secrets during a build
1. Volumes. Volumes don't work during build.
2. `ADD` them, `RUN` something, then remove the secrets. This has the unfortunate effect of the secrets existing in plaintext in the `ADD` layer. This is the case that `SQUASH` ameliorates.
3. Operate a HTTP server, poke the secrets into it, and have the build retrieve, use, and delete them in a single `RUN` command. We did this in the past, and I've heard of others doing it, but I find it rather scary.

### CODE REVIEW

**Comment by jfrazelle** [comment]
burke!!! cooooll!!!!!

**Comment by burke** [comment]
cc @jvgogh because we were chatting about weird ways to work around this at the kubernetes thing the other month.

cc also @proppy @bwilkins @dysinger @WhisperingChaos via the #8660 train

**Comment by thaJeztah** [comment]
Interesting. Should I see this as a continuation of https://github.com/docker/docker/issues/8574 (`BEGIN .. COMMIT`) - should that be closed?

Some quick thoughts;

- Are squashed layers still cached individually? This may be important information for users, because those layers (possibly containing secrets) will persist on the server
- I don't think *naming* marks is needed at this point (unless nested marks are possible); `SQUASH` without number could just squash everything since the last `MARK`; if no `MARK` exists, either bail out or squash everything since the beginning of the Dockerfile

Will something like this be problematic?:

```dockerfile

MARK a
ADD a
ENV hello=world
RUN b

MARK b
ADD c
ENV hello=galaxy

SQUASH a foo
RUN echo $hello 
SQUASH b bar
```

Basically; nested squashes, and, not really sure how `ENV` is handled (not given that much thought yet as to what the *expected* result should be)

**Comment by burke** [comment]
Good points; I think it makes sense to remove the first argument from both commands so that:

1. Only one mark is ever active
2. `SQUASH` must be called after `MARK` -- no more integer argument option.

As for the rest:

* The squashed layers are still present on the server doing the building. For my purposes at least, this is acceptable because it's kind of taken for axiomatic that that server has access to generate secrets.

What's actually going on under the hood is that all the layers that *will* be squashed are generated completely normally, then the squash does exactly the same thing as a git rebase -- it creates a new layer (using the last image ID, the description, and the number of layers squashed as the cache key) and orphan, but does not immediately nuke, all the squashed layers. This way, future builds can still take advantage of the cache. when re-building those orphaned layers.

`ENV` is handled exactly as it should be, I think. The `Config` is just copied from the last image pre-squash to the new squashed image, which will necessarily include all `ENV`, etc. 

Your multi-mark example looks like it might break `ENV`, but it would actually just squash away a bunch of layers, or error, or something. I don't know. I'll just remove mark names.

Thanks for the feedback!

**Comment by thaJeztah** [comment]
> For my purposes at least, this is acceptable because it's kind of taken for axiomatic that that server has access to generate secrets

Yes, I think that's reasonable; but it should be documented somewhere so that users are *aware* that this is the case.

> I'll just remove mark names.

+1; they can always be added in future, if the need arises

Thanks for creating this!

**Comment by burke** [comment]
`MARK` names are now completely gone. `MARK` takes no arguments, and `SQUASH` will always and only squash back to the most recent `MARK`.

**Comment by WhisperingChaos** [comment]
@burke 

I like this feature as an optimization!  I also appreciate the observations concerning SQASH's effect on an image's history and the exposure of a secret within the committed container (image cache) on the build host.

However, the first two justifications arguing for SQUASH's inclusion: to eliminate secrets and build-time resources from the run-time image, aren't durable ones.  It would be preferable for transient files and the processes that act upon them to never have been included in the resulting run-time image nor directly affected its environment.  These reasons represent symptoms that are better addressed by isolating build-time and run-time concerns using mechanisms that don't rely on "sanitizing" the resultant image, as it's difficult to ensure complete removal of contaminates.

Also, for secrets in particular, additional safeguards should be implemented to protect their visibility/accessibility by build-time processes.  Since SQUASH doesn't affect the visibility of a secret, available through the build context, the secret is accessible to all processes executed by the Dockerfile.  Therefore, a secret can be consumed by any of these processes and made public.  This issue certainly restrains the desire to reuse Docker Hub images that employ Dockerfile commands like "ADD ." especially when attempting to build components requiring secrets, as the secret will be silently copied by the "ADD ." command.  Here's a reference to a detailed explanation of this [vulnerability](https://github.com/docker/docker/issues/7149#issuecomment-84171831).

I hope you're not surprised by my position, as even visiting only the locomotive cab of the #8660 train would suggest the assessment above.

There has been an effort to permit the inclusion of secrets via build time environment variable assignment.  See pull request #9176.

Finally, I sympathize with the desire to resolve issues, like build-time secrets, that resulted in the invention of SQUASH. Without a viable alternative, it might be prudent, especially in the short term, to ignore my position.

**Comment by burke** [comment]
Yeah, I completely agree with all of your concerns. I don't think `SQUASH` adequately solves the problem, but it was the closest I could get without performing much more major surgery on the builder.

Ultimately I think there needs to be a much better solution than this, but I figured some people might find my proposal useful regardless.

**Comment by rdsubhas** [comment]
More than the secrets part, I'm more excited about reducing container sizes! IMHO the greatest problem I've had with Dockerfiles so far is the inability to control when to cache and when not to. I frequently find a group of commands that are to be considered as **one logical group**, e.g:

* ADD Gemfile, ADD Gemfile.lock, RUN bundle install
* ADD <files>, RUN chown (the infamous user/group issue #9934)
* ADD <template>, RUN <process template>

Layering them just adds to the constant annoyance of keeping a tab on layer sizes. Many times, I find myself fighting against the Dockerfile cache than with it.

With this feature, I won't even need the (painful) upcoming syntax for COPY (COPY --user=xyz <files>) and can just use MARK and SQUASH! Yay!

I also dig the messages for MARK/SQUASH, it helps with the logical unit part. But as long as I get a way to control the layers, I can definitely live without it. Thanks a bunch @burke and hope this sees light soon...

**Comment by LK4D4** [comment]
I'm definetely +1 to this, because it allows you to do COPY + RUN in one layer.
ping @docker/core-maintainers

**Comment by TomasTomecek** [comment]
It would be really awesome if this got merged. Precisely controlling layers within dockerfile is something what we really miss (and are trying to workaround it).

**Comment by duglin** [comment]
I need to find some time to dig more into this, in particular around the caching, but it sounds promising. One question, would it make sense to have a `SQUASH` w/o a preceding `MARK` squash all layers down to one?

**Comment by thaJeztah** [comment]
> One question, would it make sense to have a SQUASH w/o a preceding MARK squash all layers down to one?

Good point; would it gain much over putting `MARK` directly after the `FROM`?

**Comment by duglin** [comment]
Just a nice shortcut. I figured squashing down to one layer is probably one of the more common usecases.

**Comment by thaJeztah** [comment]
True, otoh, it would mask cases where someone forgot to put a `MARK` in. We could leave it as a future enhancement; IIC, that won't be a breaking change.

(Heh, thinking, people would probably try it the hackish way; `ONBUILD MARK` :))

**Comment by burke** [comment]
I think I made `ONBUILD` explicitly disallow `MARK`. `SQUASH` and `MARK` are implemented a little differently than other commands, as they're just metadata on the `builder`, rather than an actual in-tree representation. While I'm sure it could be made to work with `ONBUILD` somehow, I'd rather this be as simple and explicit as possible.

**Comment by burke** [comment]
So far as caching, please do look into it, but the way it works with this patch is roughly:

1. Each layer after the mark is built exactly as before, with caching carried out as before.
2. Upon a `SQUASH`, all the intermediate layers are orphaned, as the current `builder` is rewound to the layer immediately before the `MARK`
3. A new layer is created on top with:
  * Contents as the diff between the new top layer and the old top layer
  * Metadata of the old top layer, except:
  * `ContainerConfig.Cmd` is changed to provide a readable `docker history`

Caching for this squashed layer is keyed on the parent image ID (the last pre-squash image ID), the number of layers to squash, and the replacement `docker history` message.

**Comment by burke** [comment]
Downsides I can think of off the top of my head:

1. Anyone running GC scripts that delete orphaned images is going to have a bad time wrt caching of builds using squashing.
2. Not only is this not helpful in improving build time as a true `BEGIN/COMMIT` sort of scheme would be, it's actually somewhat harmful, as calculating the diff between the two layers requires a tar/untar, which can be expensive on large images.
3. If people are putting very secret things inside, it will be somewhat hard to properly educate them that the images really do stick around on disk (though they won't be pushed, and it's a safe assumption that any secrets are already accessible given local `docker` group access.

**Comment by thaJeztah** [comment]
The `ONBUILD` was just something people would possibly try, but I fully agree it shouldn't be supported, because it doesn't really make sense, so :+1:

**Comment by TomasTomecek** [comment]
> would it make sense to have a SQUASH w/o a preceding MARK squash all layers down to one?

I think you probably want to have at least 2 layers:
1. Your newly built image 
2. Base image 

What's the reasoning to have newly built image merged into base image?

**Comment by phemmer** [comment]
> What's the reasoning to have newly built image merged into base image?

You might use a base image with development tools, build a static binary, and then wipe the entire filesystem except for your binary.

**Comment by Sirupsen** [comment]
My concern with introducing this syntax is that Dockerfiles now leaks an abstraction previously hidden from the user. Dockerfiles and the build system need a fundamental overhaul and this feels like a patchy solution on top of a system that doesn't accommodate complex scenarios. Dockerfiles are easy to use, but they don't sit at a layer where they enable enough extensibility and control. I know there's talk of splitting Dockerfiles from the build primitives, and ultimately I think that's the right approach. That said, I'm concerned with how long that could take and this helps enable some shortcomings for advanced users.

**Comment by tiborvass** [comment]
I 100% share @Sirupsen's concerns.

**Comment by burke** [comment]
Me too actually. I proposed this just because it's easy and the current situation is inadequate.

**Comment by graemej** [comment]
I'd argue that the layers abstraction has already leaked in that `Dockerfile` consumers need to consciously minimize the number of layers (per [best-practices doc](https://docs.docker.com/articles/dockerfile_best-practices/#minimize-the-number-of-layers)) and the practice of chaining commands (`&&`) is widespread.

Seems to me that a bit of layer control allows `Dockerfiles` to address a wider range of use cases and could save people from creating customized build systems.  I'd also love to see a stronger set of build primitives, but could live with a better `Dockerfile` in the interim.

**Comment by rdsubhas** [comment]
+1 to @graemej. With Docker, I really don't think layering is an internal detail, its a fundamental feature that we consider and work with all the time. To my (*very limited*) knowledge leaky abstractions are for internal details that are hidden from the user but accidentally exposed, but here I wonder if layering is something hidden at all in the first place.

To me, this feature is the same as `docker run` with `--lxc-conf`. Yeah one can run Docker without even knowing its all namespaces/cgroups/etc behind the scenes, but giving them an opportunity to tweak is not really a leaky abstraction when that's a core feature.

Maybe whether Dockerfile is the right place or not could be a contention, but whether Layering and Caching is a "hidden implementation detail" - I'm really not sure. It feels to me that even if build primitives are separated away from Dockerfile in a far future, there would still be a need to control layering wherever build happens.

**Comment by Sirupsen** [comment]
Really good point @graemej :+1:  Did not know about that documentation (which also doesn't clarify why you'd want to limit layers and how to make efficient layers, unfortunately).

**Comment by rhatdan** [comment]
We really want this patch.  Some teams packaging software are seeing huge numbers of layers, LABEL patch allow makes the situation worse.  Chainging commands together makes the Dockerfile difficult to read.  Lets get this into docker-1.7  Finally get squashing.

**Comment by cpuguy83** [comment]
I don't like anything that exposes the caching implementation in the Dockerfile. I think it's the wrong approach.

Caching exists to speed up the build at the expense of extra storage.
Images can be squashed completely separate from the build process.

I think if on push (or optionally at the end of build it would keep the cache layers but also make a squashed version?) we squash down to the nearest tagged image this ought to be good enough. This would also preserve the distribution/storage benefits that layers provide by preserving the Base layers.

**Comment by vbatts** [comment]
@cpuguy83 i would hope nothing negates caching. That is a desirable feature. The squashing should be beyond it though. If the layers produced in build still exist, and the result is squash to the `FROM`, then now the `FROM` has 2 children.
Even when we are switched to the v2 image format, layering may be abstracted, but the notion of squashing back the `FROM` is going to be needed.

**Comment by cpuguy83** [comment]
btw, my idea is basically the equivalent of doing https://gist.github.com/cpuguy83/f0b67b52537b5672c68a but without the actual syntax in the Dockerfile.
Maybe a flag or something on `docker build` that produces an additional tagged image that has been squashed.

Doesn't fix everyone's use-case, I know.

**Comment by jacobat** [comment]
I haven't read the full thread so I apologize if my comment is a bit out of context.

We have a use case where squashing layers would be pretty heavy on us.

Generally our Dockerfiles tend to have the following structure / rhythm:

```
FROM ubuntu

# Install some packages
RUN apt-get ...

# Install a bunch of libraries
COPY Gemfile /app/
COPY Gemfile.lock /app/
RUN bundle install

# Add the application code
COPY . /app/

...
```

The layers above the application code tend to change at a much slower pace than the application code itself. They also tend to be a lot larger than the application code (we're closing in on 1GB before we get to the application code). With the current layered approach when a new image is built only the application layer will need to be redistributed. If we were to squash everything we would be redistributing the same GB worth of bits.

**Comment by vbatts** [comment]
Fair. Isn't that basically what this PR does?

Also, having a first class `docker squash ...`  command could eliminate
aspects of this fro. The Dockerfile completely.

**Comment by vbatts** [comment]
Jacob, that mild misunderstanding misses the point. If you could squash
your base (just below apication code) back to the FROM, and then layer on
that, you could escrow the most common layers and have added savings.

**Comment by cpuguy83** [comment]
@vbatts Sure, except without the `MARK` and `SQUASH` snytax in the Docekrfile.

**Comment by jacobat** [comment]
@vbatts Not sure I understand. Are you saying I can squash only some of the layers in the Dockerfile?

**Comment by burke** [comment]
Yes, insofar as this PR adds the ability to squash a range of lines in a Dockerfile into a single layer (as explained in the PR description).

No one is suggesting mandatorily squashing all layers into a single layer, or doing away with the whole concept of multiple layers.

**Comment by LK4D4** [comment]
@burke Actually @cpuguy83 suggesting this indeed :)

**Comment by thaJeztah** [comment]
A squash **command** is actually what started all these... see https://github.com/docker/docker/pull/4232

**Comment by vbatts** [comment]
/me all too familiar
;-)

**Comment by burke** [comment]
@LK4D4 Sure, I should have put more emphasis on *mandatorily* -- unless you have context I don't -- it didn't read like @cpuguy83 thinks layers should go away in their entirety, just that he'd rather have an opt-in flag that compresses all layers in a build rather than a new syntax element in Dockerfiles.

This can be (rather awkwardly) composed into exactly the same feature (less the history prettification) just by chaining multiple Dockerfiles where each Dockerfile represents a MARK/SQUASH block.

Ultimately I think the problem is that the builder is too innately wired into the the-concept-that-is-Docker, and its limitations tend to become limitations of Docker. That's not a trivially-actionable though.

**Comment by WhisperingChaos** [comment]
Most COPY/ADD operations don't need to be cached, nor generate a new layer.  The resources provided by COPY/ADD  are already "cached", as they exist as part of the build context.  Also, new layer generation would be unnecessary, as long as successive COPY/ADD operators don't overwrite the same locations.  Certainly one would have to remember the checksum of COPY/ADD resources to determine if the RUN statement that depends on them requires execution (invalidates the cache produced by the RUN command).

In other words, most ADD/COPY operations could be replaced by a virtual link or system link referring to the object's location in the build context.  I would suggest something akin to the proposed SLINK #7654 instruction to reduce the size of the resulting image and improve build performance by eliminating the physical copying (replication) of the build context resources.  If the SLINK instruction reflected the checksum of the actual files it pointed to, it could seamlessly replace COPY/ADD in most situations.

Finally, many of the symptoms addressed by SQUASH are related to the separation of build-time vs run-time concerns.  If this problem were resolved, SQUASH would become less necessary, as final image production would be a series of copy operations to lift the final build artifacts from prior intermediate build containers to the resulting image, thereby minimizing layer generation.

**Comment by TomasTomecek** [comment]
@cpuguy83
> I don't like anything that exposes the caching implementation in the Dockerfile. I think it's the wrong approach.
> 
> Caching exists to speed up the build at the expense of extra storage.
> Images can be squashed completely separate from the build process.

It would be awesome if docker daemon had API endpoint for that.

> I think if on push (or optionally at the end of build it would keep the cache layers but also make a squashed version?) we squash down to the nearest tagged image this ought to be good enough. This would also preserve the distribution/storage benefits that layers provide by preserving the Base layers.

I think that overall performance is way more important than caching and/or build time. I mean, yeah, it's very nice when I'm testing some image and I can continuously rebuild it in a matter of seconds. But in the end, if my image is 1 GB big and squashed version could take only 800 MB or even 600 MB and I really need to put a semi-complex bash script inside Dockerfile to reduce the final size and number of layers, I pick performance over build time.

Also, I totally don't understand why instructions like `ENV`, `VOLUME` or `EXPOSE` create new layer. I feel like those should be completely separated from build itself. (I guess that's the same thing which was already mentioned here: build time vs. runtime)

Since cache was already mentioned: I really miss precise cache control. E.g. when I want to update all my packages in one `RUN` instruction, I would like to be sure that I get latest packages **all the time** (not cached version which is 1 month old). Controlling cache within dockerfile or via `docker build` could help, but that's certainly not ideal. (I mean that I would like to invalidate cache of some layers)

**Comment by duglin** [comment]
> On May 13, 2015, at 11:47 AM, Tomas Tomecek <notifications@github.com> wrote:
> Since cache was already mentioned: I really miss precise cache control. E.g. when I want to update all my packages in one RUN instruction, I would like to be sure that I get latest packages all the time (not cached version which is 1 month old). Controlling cache within dockerfile or via docker build could help, but that's certainly not ideal. (I mean that I would like to invalidate cache of some layers)
> 
Be careful, you may be asking for #10682 to enable turning off caching at a certain point in the Dockefile processing  :-)
-Doug

**Comment by duglin** [comment]
I really think part of the problem with issues like this is that there are some fundamental differences around what a `Dockerfile` is and what is it for.

For example, if you view a Dockerfile as the "definition" of how to build a single Docker image then it makes sense that would might push back on the notion of `NOCACHE` or `MARK`/`SQUASH` in there because you may think that the person executing the build knows what's going on, and should have total control, and the only thing you care about is the end result - meaning the single final image - so squashing outside of the build process seems perfectly reasonable.

However, if you view a Dockerfile more like a Makefile, a set of steps that you're trying to automate, then you'll want to add as many of those steps to the Dockerfile as you can because, as the author, you're the one who knows what's going on; knows what's supposed to be generated; and (let's be honest :-) ) we don't trust the person kicking off the build to remember to squash, turn off caching, etc....  so we need to be able to force those things to happen via the one common automation mechanism we have, Dockerfiles.

The point of this... until we come to an agreement on what a Dockerfile is and what's its meant to be for, I think we'll continue to frustrate people on both sides of issues like this.

It might be helpful if some old-timers on the project ( e.g. @shykes @crosbymichael etc...) could help set some direction on this topic.

Personally, I view it more like a Makefile and want to push as much as possible into it so that the person running the build needs to do (and remember to do) as little as possible.  If they really don't like what's in there, they can fork the Dockerfile.

**Comment by rdsubhas** [comment]
@duglin well put :+1:

**Comment by rhatdan** [comment]
People have been asking for this for well over a year.  Everytime we get a good solution, someone comes in an derails it.  Why can't we have both docker squash and MARK/SQUASH in Dockerfile.  I am tired of waiting for the "promised land", lets give users/developers something to fix this problem, and continue working on breaking Dockerfile functionality into base primitives.

**Comment by cpuguy83** [comment]
Closing due to comment from https://github.com/docker/docker/issues/332#issuecomment-102133413

>Review with @duglin @crosbymichael @diogomonica @LK4D4
>
>The problem with this issue is that it provides a solution to a problem that yet has to be defined
>
>The bigger question is what is the end goal for users who need to squash layers? Faster pull/push? Removing some layers (with secrets)? Maximum number of layers?
>
>It feels like all these problems are implementation details and solving them by exposing squashing is >probably the wrong way.

>We suggest to address each independent issue you guys have, in more focused separate issues.
>
>For relieving the immediate pain, we also suggest updating the documentation on how to manually squash layers if people really need it, with external tools. @crosbymichael volunteered on writing such a tool.
>
>We're closing this issue. Would love to continue the debate on more focused issues.

**Comment by burke** [comment]
> We suggest to address each independent issue you guys have, in more focused separate issues.

This is fine (and IMO the correct course), but I'd like to point out that this has been the official position for 1.5 years.

**Comment by TomasTomecek** [comment]
> The problem with this issue is that it provides a solution to a problem that yet has to be defined

I think it can be defined very easily:

 * we can't remove "secrets" from final image

 * we can't precisely control number of layers

> Faster pull/push?

I personally don't care about this but can imagine that saving bandwidth can be very important for someone.

> Removing some layers (with secrets)?

Yes.

> Maximum number of layers?

Yes (due to performance).

> We're closing this issue. Would love to continue the debate on more focused issues.

Please, provide a link in this issue to the debate.

**Comment by thaJeztah** [comment]
> We suggest to address each independent issue you guys have, in more focused separate issues.

Ok, reasonable, I guess

> For relieving the immediate pain, we also suggest updating the documentation on how to manually squash layers if people really need it, with external tools. @crosbymichael volunteered on writing such a tool.

Obviously, this won't allow optimised images when using automated builds on Docker Hub. Which makes it not very useful, or, just as useful as creating an image interactively and committing the result

**Comment by tjdett** [comment]
> The bigger question is what is the end goal for users who need to squash layers? Faster pull/push? Removing some layers (with secrets)? Maximum number of layers?

> We suggest to address each independent issue you guys have, in more focused separate issues.

Excellent. My single, focused issue is reducing the overall size of images to preserve disk space on  servers. Reducing total bytes transferred is useful, but secondary.

My standard Dockerfile development process tends towards starting with multiple run statements to gain the benefits of caching on slow operations:

```Dockerfile
RUN git checkout https://github.com/foo/bar.git
RUN cd /bar && make && make install && cd ..
RUN rm -rf bar
```

Then when the build succeeds, I need to manually coalesce and rebuild to avoid including source files and intermediate build files in the image:

```Dockerfile
RUN git checkout https://github.com/foo/bar.git && \
  cd /bar && make && make install && cd .. && \
  rm -rf bar
```

This is obviously more work for me with no obvious gain, so either there's a problem in my methodology or the tool. This PR removes the extra work, with the SQUASH comment handily able to supplement my existing Dockerfile commenting.

> It feels like all these problems are implementation details and solving them by exposing squashing is probably the wrong way.

Much like cloned git repos are always bigger than their archives for any non-trivial project, you can't have both a version-based layering system and the smallest image possible for its contents. The extra information has to be stored somehow, even if it has no practical use in the context you're using it.

If the intermediate steps serve no purpose to the final product, why insist on them being kept to gain the benefits of build caching? The same applies for Git pull requests. You submit squashed commits based on discrete chunks of functionality, not based on the writing process that produced the code, so you have a more useful historical record, rather than a more complete but noisy one.

> For relieving the immediate pain, we also suggest updating the documentation on how to manually squash layers if people really need it, with external tools.

This is lovely, except that you can't have a Docker Hub trusted build and do this unless the squashing tool is also going to be integrated with Docker Hub. Merging this PR would improve transparency of Docker Hub images compared with the suggested fix, and Docker Hub doesn't benefit from build caching anyway.

There are other Docker registry providers who provide squashing functionality already, so this may be the more appropriate way to resolve if layering is intended to be an implementation detail of the Dockerfile builds as opposed to an integral part (which is the way lots of maintainers treat it today). If so, where should I log the feature request?

**Comment by vincentwoo** [comment]
I, too, agree with @tjdett on his ails. I also think it a bit odd that very long, storied threads like this are being closed with no link to a place to continue discussion. It's been ages now, where are the links? Would love to gripe in a more focused thread.

**Comment by t128** [comment]
Any news on this issue? Maybe a link to the focused discussion @cpuguy83 ?

**Comment by cpuguy83** [comment]
@t128 This issue is closed and won't be supported.
Layers are an implementation detail, the user should not even know about them -- the fact that they do is a bug.

In master currently (and what will be Docker 1.10) the concept of a layer will be gone from the user perspective, though still there in implementation.
Instead of having images made up of a bunch of layers, you just have images.
Images can still share layers but to the user it's just an image, there is no parent image or layer. You can't download a specific layer only an image.

**Comment by phemmer** [comment]
@cpuguy83 Is there a discussion somewhere on the subject?
I strongly disagree with hiding the concept of layers from users. There is a very significant difference in a `Dockerfile` between:

    RUN something_which_generates_temp_files
    RUN rm /path_to_temp_files

and

    RUN something_which_generates_temp_files && rm /path_to_temp_files

Until the distinction between those 2 dockerfiles goes away, users need to be aware of how their images are built.

**Comment by duglin** [comment]
@cpuguy83 you've mention this on IRC too and I keep meaning to ask you about this. How are the layers gone? When I build an image I still the layers in `docker history` and I still see the "parent" field in docker inspect` for the images/layers.  What appears to be different is that they're now sha's instead of just random IDs.  When I walk the history to an "old" image (like ubuntu) things get a bit weird, but I assumed this was because of the migration stuff, but for "new" images it sure looks like the old stuff just with new types of IDs.  Can you elaborate on how things are different?

**Comment by wmark** [comment]
@burke Thanks a lot for your work on this feature! Are you committed to maintain an up-to-date fork? Is there another approach which superseded your efforts?

**Comment by burke** [comment]
@wmark nope, sorry -- we work around this at a deeper level now by not using the builder in the first place.

**Comment by cpuguy83** [comment]
@duglin For instance, `docker images -a` does not yield any untagged layers as it used to.
The only images that have a populated "parent" field (apparently... no pun intended?) are from images you build locally.

@phemmer This is still available in `docker history`.

**Comment by rhatdan** [comment]
I guess the question though is there a difference now between
```
RUN something_which_generates_temp_files
RUN rm /path_to_temp_files
```

and
```
RUN something_which_generates_temp_files && rm /path_to_temp_files
```
Are the temp files still present in the image somewhere?

**Comment by cpuguy83** [comment]
@rhatdan The fact that engine still use layers is unchanged and as such the resulting fs is also still the same.

**Comment by rhatdan** [comment]
That is what I figured and the reason people still want squash.

**Comment by t128** [comment]
the new "solution" is to hide the fact from the user why his container size is exploding?

**Comment by thaJeztah** [comment]
@t128 you can read about some of the reasons for this change here; https://github.com/docker/docker/pull/17924, here https://gist.github.com/aaronlehmann/b42a2eaf633fc949f93b#new-image-config, and https://github.com/docker/docker/pull/18378#discussion_r46632479
